File: src/goblinmode/gpu.py
"""Deep NVIDIA GPU state probe - the stuff that explains a frame-rate cliff
that isn't thermal.

Every field here is queried straight from ``nvidia-smi``. The point is to snapshot
the GPU *at the moment of a stall* and again *after the game exits*, so an
extreme FPS dip can be pinned on a concrete cause:

* VRAM exhaustion -> the driver falls back to system memory over PCIe (huge stalls)
* PCIe link down-training -> Gen1 / narrow width = bandwidth starvation
* GPU stuck in a low power-state / collapsed core clock under load
* VRAM not released after the game exits -> a driver-side leak (needs a reboot)
"""

from __future__ import annotations

import logging
import shutil
import subprocess
import threading
import time

log = logging.getLogger(__name__)

_QUERY = (
    "utilization.gpu,utilization.memory,memory.used,memory.total,memory.free,"
    "clocks.current.graphics,clocks.max.graphics,clocks.current.memory,"
    "clocks.max.memory,pstate,pcie.link.gen.current,pcie.link.gen.max,"
    "pcie.link.width.current,pcie.link.width.max,temperature.gpu,"
    "power.draw,power.limit,clocks_event_reasons.active"
)
_FIELDS = [
    "util_gpu", "util_mem", "vram_used_mb", "vram_total_mb", "vram_free_mb",
    "clock_gfx_mhz", "clock_gfx_max_mhz", "clock_mem_mhz", "clock_mem_max_mhz",
    "pstate", "pcie_gen", "pcie_gen_max", "pcie_width", "pcie_width_max",
    "temp_c", "power_w", "power_limit_w", "event_reasons",
]


def available() -> bool:
    return shutil.which("nvidia-smi") is not None


def nvidia_module_state() -> dict:
    """Read-only: the current ``nvidia_drm.modeset`` parameter and, if
    present, the driver's reported GSP firmware version - both informational,
    neither is writable at runtime (modeset is a boot-time modprobe option;
    see ``helper.set_nvidia_modeset`` for changing the on-disk config)."""
    from pathlib import Path

    modeset_path = Path("/sys/module/nvidia_drm/parameters/modeset")
    try:
        modeset = modeset_path.read_text().strip() or None
    except OSError:
        modeset = None

    gsp_version = None
    gpus_dir = Path("/proc/driver/nvidia/gpus")
    try:
        for gpu_dir in gpus_dir.iterdir():
            info = (gpu_dir / "information").read_text()
            for line in info.splitlines():
                if line.lower().startswith("gsp firmware version"):
                    gsp_version = line.split(":", 1)[-1].strip()
                    break
            if gsp_version:
                break
    except OSError:
        pass

    return {
        "present": modeset_path.exists() or gpus_dir.exists(),
        "modeset": modeset,          # "Y" / "N" / None (unreadable or no nvidia_drm)
        "gsp_firmware_version": gsp_version,  # None means GSP is off or unreported
    }


def _num(v: str):
    v = v.strip()
    if v in ("", "[N/A]", "N/A", "[Not Supported]"):
        return None
    try:
        return int(v)
    except ValueError:
        try:
            return float(v)
        except ValueError:
            return v


_LIGHT_QUERY = "utilization.gpu,temperature.gpu,clocks_event_reasons.active"


def deep_state() -> dict:
    """A full GPU snapshot. Runs two ``nvidia-smi`` subprocesses (~0.1-1 s, but
    occasionally seconds under load) - **never call this from a GLib main loop**;
    use :class:`GpuMonitor` for the polled case."""
    if not available():
        return {}
    try:
        out = subprocess.run(
            ["nvidia-smi", f"--query-gpu={_QUERY}", "--format=csv,noheader,nounits"],
            capture_output=True, text=True, timeout=5,
        ).stdout.strip().splitlines()
    except (OSError, subprocess.SubprocessError) as exc:
        log.warning("nvidia-smi deep query failed: %s", exc)
        return {}
    if not out:
        return {}
    parts = [p.strip() for p in out[0].split(",")]
    state = {k: _num(parts[i]) for i, k in enumerate(_FIELDS) if i < len(parts)}
    state.update(_pcie_throughput())
    return state


def light_state() -> tuple[float | None, float | None, str]:
    """(gpu load %, temp °C, clock-event-reasons hex) - one cheap nvidia-smi."""
    if not available():
        return None, None, ""
    try:
        out = subprocess.run(
            ["nvidia-smi", f"--query-gpu={_LIGHT_QUERY}", "--format=csv,noheader,nounits"],
            capture_output=True, text=True, timeout=4,
        ).stdout.strip().splitlines()
    except (OSError, subprocess.SubprocessError):
        return None, None, ""
    if not out:
        return None, None, ""
    parts = [p.strip() for p in out[0].split(",")]
    try:
        return float(parts[0]), float(parts[1]), (parts[2] if len(parts) > 2 else "")
    except (ValueError, IndexError):
        return None, None, ""


class GpuMonitor:
    """Polls ``nvidia-smi`` on a background thread so the daemon's GLib loop
    never blocks on it. Everything degrades to empty when there is no NVIDIA GPU.

    * :meth:`light` - (load, temp, reasons), refreshed roughly every second
    * :meth:`deep`  - the full snapshot dict, refreshed roughly every 8 s
    """

    #: (light interval, deep interval) seconds - idle first, then while a game runs
    _IDLE = (6.0, 30.0)
    _ACTIVE = (1.0, 5.0)

    def __init__(self, deep_interval: float | None = None) -> None:
        self._active = False
        self._active_deep = deep_interval or self._ACTIVE[1]
        self._lock = threading.Lock()
        self._light: tuple[float | None, float | None, str] = (None, None, "")
        self._deep: dict = {}
        self._deep_at = 0.0
        self._thread: threading.Thread | None = None
        self._stop = threading.Event()
        self._wake = threading.Event()

    def start(self) -> None:
        if not available() or (self._thread and self._thread.is_alive()):
            return
        self._stop.clear()
        self._thread = threading.Thread(target=self._run, name="gmp-gpu", daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        self._wake.set()

    def set_active(self, active: bool) -> None:
        """Poll fast while a game runs, slowly otherwise."""
        if active != self._active:
            self._active = active
            self._wake.set()  # apply the new cadence immediately

    def _run(self) -> None:
        while not self._stop.is_set():
            light_iv, deep_iv = self._ACTIVE if self._active else self._IDLE
            deep_iv = self._active_deep if self._active else deep_iv
            light = light_state()                      # subprocess - outside the lock
            with self._lock:
                self._light = light
            if time.monotonic() - self._deep_at >= deep_iv:
                deep = deep_state()                   # subprocess - outside the lock
                with self._lock:
                    self._deep = deep
                    self._deep_at = time.monotonic()
            self._wake.wait(light_iv)
            self._wake.clear()

    def light(self) -> tuple[float | None, float | None, str]:
        with self._lock:
            return self._light

    def deep(self, *, force: bool = False) -> dict:
        """Cached deep snapshot. ``force`` blocks for a fresh one - only call
        that from a worker thread (e.g. the FPS-dip handler)."""
        if force:
            fresh = deep_state()
            with self._lock:
                self._deep = fresh
                self._deep_at = time.monotonic()
            return fresh
        with self._lock:
            return dict(self._deep)


def _pcie_throughput() -> dict:
    """rx/tx PCIe throughput in MB/s from `nvidia-smi dmon` (one sample)."""
    if not available():
        return {}
    try:
        rows = subprocess.run(
            ["nvidia-smi", "dmon", "-s", "t", "-c", "1"],
            capture_output=True, text=True, timeout=5,
        ).stdout.strip().splitlines()
    except (OSError, subprocess.SubprocessError):
        return {}
    for row in rows:
        row = row.strip()
        if not row or row.startswith("#"):
            continue
        cols = row.split()
        # "# Idx  rxpci  txpci" -> idx, rx, tx  (units MB/s)
        if len(cols) >= 3:
            return {"pcie_rx_mbps": _num(cols[1]), "pcie_tx_mbps": _num(cols[2])}
    return {}


def classify_dip(state: dict, cpu_load: float | None, disk_read_mbps: float | None) -> str | None:
    """When FPS collapses but nothing is working hard, the frames are being
    *withheld* (focus loss / loading screen), not *starved* (a bottleneck).
    Return a plain-language note if that's what this looks like, else None.
    """
    util = state.get("util_gpu")
    if util is None:
        return None
    gpu_idle = util < 15
    cpu_light = cpu_load is None or cpu_load < 45
    if gpu_idle and cpu_light:
        if disk_read_mbps and disk_read_mbps > 25:
            return (
                f"CPU and GPU were near-idle while the disk read {disk_read_mbps:.0f} MB/s "
                f"- this is a loading screen / zone transition streaming assets, not a bottleneck"
            )
        return (
            "CPU and GPU were both near-idle - the frames were withheld, not starved. "
            "Usually the game window losing focus (alt-tab -> WoW 'Max Background FPS' "
            "cap), a loading screen, or a menu. Not a hardware problem"
        )
    return None


def assess(state: dict, *, fps: float | None = None, under_load: bool = True) -> list[str]:
    """Return human-readable likely causes for a stall, most-damning first.

    ``under_load`` should reflect whether the GPU/CPU were actually busy - pass
    False and the PCIe / power-state / clock heuristics are skipped, because a
    down-trained link or a low P-state is *expected* when the GPU is idle.
    """
    if not state:
        return []
    out: list[str] = []

    # a real load check from the state itself, so a stale/optimistic under_load
    # can't produce false "bandwidth-starved" lines while the GPU sleeps
    util = state.get("util_gpu") or 0
    busy = under_load and util >= 25

    used, total = state.get("vram_used_mb"), state.get("vram_total_mb")
    free = state.get("vram_free_mb")
    if used and total:
        frac = used / total
        if frac >= 0.94 or (free is not None and free < 300):
            out.append(
                f"VRAM near exhaustion ({used}/{total} MB, {free} MB free) - the "
                f"driver is likely spilling to system RAM over PCIe"
            )
        elif frac >= 0.88:
            out.append(f"VRAM pressure high ({used}/{total} MB)")

    rx = state.get("pcie_rx_mbps")
    gen, gen_max = state.get("pcie_gen"), state.get("pcie_gen_max")
    w, w_max = state.get("pcie_width"), state.get("pcie_width_max")
    # only a bottleneck if the link is BOTH degraded AND actually being pushed
    if busy and gen and gen_max and gen < gen_max and (rx is None or rx > 500):
        out.append(f"PCIe link at Gen{gen} (card supports Gen{gen_max}) while carrying "
                   f"{rx} MB/s - bandwidth-starved")
    if busy and w and w_max and w < w_max and w <= 4 and (rx is None or rx > 500):
        out.append(f"PCIe link narrowed to x{w} (of x{w_max}) under load")

    ps = state.get("pstate")
    if busy and ps in ("P5", "P8", "P12", "P15") and util > 50:
        out.append(f"GPU stuck in low power-state {ps} while {util}% utilised")

    cg, cgm = state.get("clock_gfx_mhz"), state.get("clock_gfx_max_mhz")
    if busy and cg and cgm and cg < cgm * 0.55 and util > 50:
        out.append(f"GPU core clock collapsed to {cg}/{cgm} MHz under {util}% load")

    er = state.get("event_reasons")
    if isinstance(er, int) and er not in (0,):
        bad = {0x8: "HW slowdown", 0x40: "HW thermal", 0x80: "HW power-brake"}
        hit = [v for k, v in bad.items() if er & k]
        if hit:
            out.append("nvidia clock-event: " + ", ".join(hit))

    return out


def describe_dip(
    state: dict,
    *,
    fps: float,
    baseline: float,
    cpu_load: float | None,
    disk_read: float | None,
    cpu_core_max: float | None = None,
) -> tuple[str, bool]:
    """Turn a fresh GPU snapshot + the dip numbers into an incident line.

    Returns ``(detail, is_real)``. ``is_real`` is False for a dip that isn't a
    hardware *fault* - frames withheld (focus loss / loading), or a scene that's
    simply GPU- or CPU-bound at the current settings - so it doesn't arm the
    post-game post-mortem. Mutates ``state`` with the at-dip context the
    exporter and Diagnostics page read back.
    """
    gpu_busy = (state.get("util_gpu") or 0) >= 25 or (cpu_load or 0) >= 60
    benign = classify_dip(state, cpu_load, disk_read)
    causes = assess(state, fps=fps, under_load=gpu_busy)
    state["likely_causes"] = causes
    state["cpu_load_at_dip"] = round(cpu_load, 1) if cpu_load is not None else None
    state["cpu_core_max_at_dip"] = round(cpu_core_max, 1) if cpu_core_max is not None else None
    state["disk_read_mbps_at_dip"] = disk_read

    util = state.get("util_gpu")
    at = f"(baseline ~{baseline:.0f})"

    if benign and not causes:
        state["assessment"] = "benign - not a hardware bottleneck"
        return f"Frame rate dipped to {fps:.0f} FPS {at}. {benign}", False
    if causes:
        return f"Frame rate collapsed to {fps:.0f} FPS {at}. {causes[0]}", True

    dropped = baseline > 0 and fps <= baseline * 0.75
    if dropped and util is not None and util >= 92:
        state["assessment"] = "GPU-bound scene"
        return (
            f"Frame rate dropped to {fps:.0f} FPS {at}. GPU pegged at {util:.0f}% - "
            f"this spot is heavier than your settings can sustain, not a fault. "
            f"Lower a setting or cap the frame rate here.",
            False,
        )
    if dropped and util is not None and util < 80 and cpu_core_max is not None and cpu_core_max >= 95:
        state["assessment"] = "CPU-bound scene"
        return (
            f"Frame rate dropped to {fps:.0f} FPS {at}. A CPU core was pegged at "
            f"{cpu_core_max:.0f}% while the GPU had headroom ({util:.0f}%) - a "
            f"single-threaded hotspot (busy city, raid), not a hardware fault.",
            False,
        )
    return (
        f"Frame rate dropped to {fps:.0f} FPS {at}. No single cause stood out - "
        f"the GPU snapshot is attached. A short drop like this is usually a zone "
        f"load, shader compilation or a background task.",
        True,
    )


def post_mortem(idle_state: dict) -> tuple[str, str] | None:
    """After the game exits: did the GPU actually let go? Returns (kind, detail).

    Only VRAM is checked - an idle PCIe down-train is normal ASPM, not a fault.
    """
    used = idle_state.get("vram_used_mb")
    if used is not None and used > 900:
        return (
            "vram_not_freed",
            f"{used} MB of VRAM still allocated after the game exited - a "
            f"driver-side leak; a reboot clears it",
        )
    return None

File: crates/gmp-core/src/gpu.rs
//! Reading a GPU snapshot and saying what a frame-rate dip probably was.
//!
//! A port of the pure half of `src/goblinmode/gpu.py`. Everything that talks
//! to nvidia-smi or sysfs - `available`, `deep_state`, `light_state`,
//! `nvidia_module_state`, `GpuMonitor` - stays in Python. What moves here is
//! the judgement: given a snapshot and some numbers, is this a hardware fault,
//! a scene that is simply heavy, or the game not drawing at all?
//!
//! The distinction matters more than it sounds. A dip that is not a fault must
//! not arm the post-game post-mortem, or the user is told something broke
//! every time they alt-tab.

use serde_json::{Map, Value};

use crate::round::one_dp;

/// A GPU snapshot: the heterogeneous dict nvidia-smi parsing produces.
pub type State = Map<String, Value>;

/// A numeric field, or None if absent, null or not a number.
fn num(state: &State, key: &str) -> Option<f64> {
    state.get(key).and_then(Value::as_f64)
}

/// A numeric field that is present AND non-zero.
///
/// Python spells this `if used and total`, where 0 is falsy. Translating that
/// as "is present" would make a zeroed field look like a real reading.
fn nonzero(state: &State, key: &str) -> Option<f64> {
    num(state, key).filter(|v| *v != 0.0)
}

/// Parse one nvidia-smi cell.
///
/// The sentinels are nvidia-smi's own way of saying a field does not apply to
/// this card. A value it does not recognise comes back as the raw string
/// rather than being dropped, so an unexpected format is visible in a bug
/// report instead of silently becoming null.
pub fn parse_cell(v: &str) -> Value {
    let v = v.trim();
    if matches!(v, "" | "[N/A]" | "N/A" | "[Not Supported]") {
        return Value::Null;
    }
    if let Ok(i) = v.parse::<i64>() {
        return Value::from(i);
    }
    if let Ok(f) = v.parse::<f64>() {
        return Value::from(f);
    }
    Value::from(v)
}

/// When FPS collapses but nothing is working hard, the frames are being
/// *withheld* rather than *starved*. Returns a plain-language note if so.
pub fn classify_dip(
    state: &State,
    cpu_load: Option<f64>,
    disk_read_mbps: Option<f64>,
) -> Option<String> {
    let util = num(state, "util_gpu")?;
    let gpu_idle = util < 15.0;
    let cpu_light = cpu_load.is_none_or(|c| c < 45.0);
    if !(gpu_idle && cpu_light) {
        return None;
    }
    if let Some(disk) = disk_read_mbps {
        if disk > 25.0 {
            return Some(format!(
                "CPU and GPU were near-idle while the disk read {disk:.0} MB/s \
                 - this is a loading screen / zone transition streaming assets, not a bottleneck"
            ));
        }
    }
    Some(
        "CPU and GPU were both near-idle - the frames were withheld, not starved. \
         Usually the game window losing focus (alt-tab -> WoW 'Max Background FPS' \
         cap), a loading screen, or a menu. Not a hardware problem"
            .to_owned(),
    )
}

/// Likely causes for a stall, most damning first.
///
/// `under_load` should reflect whether the GPU was actually busy. A
/// down-trained PCIe link or a low power state is EXPECTED on an idle GPU, so
/// reporting them then would be a false alarm every time.
pub fn assess(state: &State, under_load: bool) -> Vec<String> {
    if state.is_empty() {
        return Vec::new();
    }
    let mut out = Vec::new();

    // Checked against the snapshot itself, so a stale or optimistic
    // `under_load` cannot produce "bandwidth-starved" lines while the GPU sleeps.
    let util = num(state, "util_gpu").unwrap_or(0.0);
    let busy = under_load && util >= 25.0;

    let used = nonzero(state, "vram_used_mb");
    let total = nonzero(state, "vram_total_mb");
    let free = num(state, "vram_free_mb");
    if let (Some(used), Some(total)) = (used, total) {
        let frac = used / total;
        let free_txt = free.map_or("null".to_owned(), fmt_num);
        if frac >= 0.94 || free.is_some_and(|f| f < 300.0) {
            out.push(format!(
                "VRAM near exhaustion ({}/{} MB, {free_txt} MB free) - the \
                 driver is likely spilling to system RAM over PCIe",
                fmt_num(used),
                fmt_num(total)
            ));
        } else if frac >= 0.88 {
            out.push(format!(
                "VRAM pressure high ({}/{} MB)",
                fmt_num(used),
                fmt_num(total)
            ));
        }
    }

    let rx = num(state, "pcie_rx_mbps");
    let pushed = rx.is_none_or(|r| r > 500.0);
    let rx_txt = rx.map_or("None".to_owned(), fmt_num);
    if let (Some(gen), Some(gen_max)) = (nonzero(state, "pcie_gen"), nonzero(state, "pcie_gen_max"))
    {
        if busy && gen < gen_max && pushed {
            out.push(format!(
                "PCIe link at Gen{} (card supports Gen{}) while carrying {rx_txt} MB/s - bandwidth-starved",
                fmt_num(gen),
                fmt_num(gen_max)
            ));
        }
    }
    if let (Some(w), Some(w_max)) = (
        nonzero(state, "pcie_width"),
        nonzero(state, "pcie_width_max"),
    ) {
        if busy && w < w_max && w <= 4.0 && pushed {
            out.push(format!(
                "PCIe link narrowed to x{} (of x{}) under load",
                fmt_num(w),
                fmt_num(w_max)
            ));
        }
    }

    if busy && util > 50.0 {
        if let Some(ps) = state.get("pstate").and_then(Value::as_str) {
            if matches!(ps, "P5" | "P8" | "P12" | "P15") {
                out.push(format!(
                    "GPU stuck in low power-state {ps} while {}% utilised",
                    fmt_num(util)
                ));
            }
        }
    }

    if let (Some(cg), Some(cgm)) = (
        nonzero(state, "clock_gfx_mhz"),
        nonzero(state, "clock_gfx_max_mhz"),
    ) {
        if busy && cg < cgm * 0.55 && util > 50.0 {
            out.push(format!(
                "GPU core clock collapsed to {}/{} MHz under {}% load",
                fmt_num(cg),
                fmt_num(cgm),
                fmt_num(util)
            ));
        }
    }

    // Only an integer counts: the field is a bitmask, and a float here would
    // mean nvidia-smi returned something unexpected.
    if let Some(er) = state.get("event_reasons").and_then(Value::as_i64) {
        if er != 0 {
            let mut hit = Vec::new();
            for (bit, label) in [
                (0x8, "HW slowdown"),
                (0x40, "HW thermal"),
                (0x80, "HW power-brake"),
            ] {
                if er & bit != 0 {
                    hit.push(label);
                }
            }
            if !hit.is_empty() {
                out.push(format!("nvidia clock-event: {}", hit.join(", ")));
            }
        }
    }
    out
}

/// Render a number the way Python's `str()` would inside an f-string: an
/// integral float prints without a trailing `.0`.
fn fmt_num(v: f64) -> String {
    if v.fract() == 0.0 && v.abs() < 1e15 {
        format!("{}", v as i64)
    } else {
        format!("{v}")
    }
}

/// Turn a fresh snapshot plus the dip numbers into an incident line.
///
/// Returns `(detail, is_real)`. `is_real` is false for a dip that is not a
/// hardware FAULT - frames withheld, or a scene simply GPU- or CPU-bound at
/// the current settings - so it does not arm the post-game post-mortem.
///
/// Mutates `state` with the at-dip context that the exporter and the
/// Diagnostics page read back out.
pub fn describe_dip(
    state: &mut State,
    fps: f64,
    baseline: f64,
    cpu_load: Option<f64>,
    disk_read: Option<f64>,
    cpu_core_max: Option<f64>,
) -> (String, bool) {
    let gpu_busy = num(state, "util_gpu").unwrap_or(0.0) >= 25.0 || cpu_load.unwrap_or(0.0) >= 60.0;
    let benign = classify_dip(state, cpu_load, disk_read);
    let causes = assess(state, gpu_busy);

    state.insert("likely_causes".into(), Value::from(causes.clone()));
    state.insert(
        "cpu_load_at_dip".into(),
        cpu_load.map_or(Value::Null, |v| Value::from(one_dp(v))),
    );
    state.insert(
        "cpu_core_max_at_dip".into(),
        cpu_core_max.map_or(Value::Null, |v| Value::from(one_dp(v))),
    );
    state.insert(
        "disk_read_mbps_at_dip".into(),
        disk_read.map_or(Value::Null, Value::from),
    );

    let util = num(state, "util_gpu");
    let at = format!("(baseline ~{baseline:.0})");

    if let Some(benign) = &benign {
        if causes.is_empty() {
            state.insert(
                "assessment".into(),
                Value::from("benign - not a hardware bottleneck"),
            );
            return (
                format!("Frame rate dipped to {fps:.0} FPS {at}. {benign}"),
                false,
            );
        }
    }
    if let Some(first) = causes.first() {
        return (
            format!("Frame rate collapsed to {fps:.0} FPS {at}. {first}"),
            true,
        );
    }

    let dropped = baseline > 0.0 && fps <= baseline * 0.75;
    if dropped {
        if let Some(util) = util {
            if util >= 92.0 {
                state.insert("assessment".into(), Value::from("GPU-bound scene"));
                return (
                    format!(
                        "Frame rate dropped to {fps:.0} FPS {at}. GPU pegged at {util:.0}% - \
                         this spot is heavier than your settings can sustain, not a fault. \
                         Lower a setting or cap the frame rate here."
                    ),
                    false,
                );
            }
            if util < 80.0 {
                if let Some(core) = cpu_core_max {
                    if core >= 95.0 {
                        state.insert("assessment".into(), Value::from("CPU-bound scene"));
                        return (
                            format!(
                                "Frame rate dropped to {fps:.0} FPS {at}. A CPU core was pegged at \
                                 {core:.0}% while the GPU had headroom ({util:.0}%) - a \
                                 single-threaded hotspot (busy city, raid), not a hardware fault."
                            ),
                            false,
                        );
                    }
                }
            }
        }
    }
    (
        format!(
            "Frame rate dropped to {fps:.0} FPS {at}. No single cause stood out - \
             the GPU snapshot is attached. A short drop like this is usually a zone \
             load, shader compilation or a background task."
        ),
        true,
    )
}

/// After the game exits: did the GPU actually let go?
///
/// Only VRAM is checked. An idle PCIe down-train is normal ASPM, not a fault,
/// and flagging it would report a problem after every session.
pub fn post_mortem(idle_state: &State) -> Option<(String, String)> {
    let used = num(idle_state, "vram_used_mb")?;
    if used > 900.0 {
        return Some((
            "vram_not_freed".to_owned(),
            format!(
                "{} MB of VRAM still allocated after the game exited - a \
                 driver-side leak; a reboot clears it",
                fmt_num(used)
            ),
        ));
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn state(pairs: Value) -> State {
        pairs.as_object().cloned().unwrap_or_default()
    }

    // ---- translated from tests/test_gpu.py -------------------------------

    #[test]
    fn vram_exhaustion_flagged() {
        let s = state(json!({"util_gpu": 90, "vram_used_mb": 5900,
                             "vram_total_mb": 6000, "vram_free_mb": 100}));
        assert!(assess(&s, true)
            .iter()
            .any(|c| c.contains("VRAM near exhaustion")));
    }

    #[test]
    fn idle_gpu_never_flags_pcie_or_pstate() {
        // A down-trained link and a low power state are both EXPECTED on an
        // idle GPU. util gates it as well as the under_load flag, so a stale
        // caller cannot produce a false alarm either.
        let s = state(json!({"util_gpu": 2, "pcie_gen": 1, "pcie_gen_max": 4, "pstate": "P8"}));
        assert!(assess(&s, false).is_empty());
        assert!(assess(&s, true).is_empty());
    }

    #[test]
    fn empty_state_returns_empty() {
        assert!(assess(&State::new(), true).is_empty());
    }

    #[test]
    fn idle_cpu_and_gpu_is_withheld_not_starved() {
        let note = classify_dip(&state(json!({"util_gpu": 3})), Some(10.0), Some(2.0)).unwrap();
        assert!(note.contains("withheld"));
    }

    #[test]
    fn busy_gpu_returns_none() {
        assert!(classify_dip(&state(json!({"util_gpu": 80})), Some(70.0), Some(0.0)).is_none());
    }

    #[test]
    fn post_mortem_flags_unreleased_vram() {
        let v = post_mortem(&state(json!({"vram_used_mb": 1500}))).unwrap();
        assert_eq!(v.0, "vram_not_freed");
    }

    #[test]
    fn post_mortem_clean_exit_returns_none() {
        assert!(post_mortem(&state(json!({"vram_used_mb": 120}))).is_none());
    }

    // ---- the port's own hazards -------------------------------------------

    #[test]
    fn a_zeroed_field_is_not_a_reading() {
        // Python spells the guard `if used and total`, where 0 is falsy.
        // Treating it as "present" would divide by zero or report a link at
        // Gen0 as degraded.
        let s = state(
            json!({"util_gpu": 90, "vram_used_mb": 0, "vram_total_mb": 0,
                             "pcie_gen": 0, "pcie_gen_max": 4}),
        );
        assert!(assess(&s, true).is_empty());
    }

    #[test]
    fn describe_dip_records_the_context_it_was_given() {
        // The exporter and the Diagnostics page read these back out, so the
        // mutation is part of the contract rather than a side effect.
        let mut s = state(json!({"util_gpu": 99}));
        let (_detail, is_real) =
            describe_dip(&mut s, 30.0, 60.0, Some(51.15), Some(4.0), Some(99.5));
        assert!(!is_real, "a GPU-bound scene is not a fault");
        assert_eq!(s["cpu_load_at_dip"], json!(51.1)); // round(x, 1), half to even
        assert_eq!(s["cpu_core_max_at_dip"], json!(99.5));
        assert_eq!(s["assessment"], "GPU-bound scene");
        assert!(s["likely_causes"].as_array().unwrap().is_empty());
    }

    #[test]
    fn a_benign_dip_does_not_arm_the_post_mortem() {
        let mut s = state(json!({"util_gpu": 3}));
        let (detail, is_real) = describe_dip(&mut s, 0.0, 0.0, Some(5.0), None, None);
        assert!(!is_real);
        assert!(detail.contains("withheld"));
    }

    #[test]
    fn parse_cell_keeps_an_unrecognised_value_rather_than_dropping_it() {
        assert_eq!(parse_cell("42"), json!(42));
        assert_eq!(parse_cell("3.5"), json!(3.5));
        assert_eq!(parse_cell("[N/A]"), Value::Null);
        assert_eq!(parse_cell("  "), Value::Null);
        // visible in a bug report instead of silently becoming null
        assert_eq!(parse_cell("something new"), json!("something new"));
    }
}
